### PR TITLE
PMM-7405 Add check conn to external exporter.

### PR DESCRIPTION
**Pull request by askomorokhov**
[PMM-7405](https://jira.percona.com/browse/PMM-7405)

- [x] https://github.com/Percona-Lab/pmm-submodules/pull/1518

Ref: 
 - https://github.com/percona/pmm-admin/pull/135
 - https://github.com/percona/pmm-managed/pull/659
 - https://github.com/percona/pmm/pull/679
 - https://github.com/percona/pmm-agent/pull/231
---
- [ ] Tests passed.
- [ ] Feature build pass.
- [ ] (Re)requested review.
- [ ] Fix conflicts with target branch.
- [ ] Update API dependency.
- [ ] Update jira ticket description if needed.
- [ ] Attach screenshots/console output to confirm new behavior to jira ticket, if applicable.

```
[root@bb074911acd2 /]# 
[root@bb074911acd2 /]# 
[root@bb074911acd2 /]# pmm-admin add external --listen-port=9121 --scheme='http' --metrics-path='metrics2' --group="redis" --service-name="my-redis22" --metrics-mode=pull
External Service added.
Service ID  : /service_id/17df7cd7-e2b1-4c5b-a126-347ba22abeb2
Service name: my-redis22
Group       : redis

[root@bb074911acd2 /]# 
[root@bb074911acd2 /]# 
[root@bb074911acd2 /]# pmm-admin add external --listen-port=9121 --scheme='http' --metrics-path='metrics2' --group="redis" --service-name="my-redis22" --metrics-mode=push
Service with name "my-redis22" already exists.
[root@bb074911acd2 /]# pmm-admin add external --listen-port=9121 --scheme='http' --metrics-path='metrics2' --group="redis" --service-name="my-redis23" --metrics-mode=push
External Service added.
Service ID  : /service_id/cd222616-3884-4f72-ab71-8e540f249415
Service name: my-redis23
Group       : redis

[root@bb074911acd2 /]# pmm-admin add external --listen-port=9121 --scheme='httpx' --metrics-path='metrics2' --group="redis" --service-name="my-redis23" --metrics-mode=push
Service with name "my-redis23" already exists.
[root@bb074911acd2 /]# pmm-admin add external --listen-port=9121 --scheme='httpx' --metrics-path='metrics2' --group="redis" --service-name="my-redis24" --metrics-mode=push
Connection check failed: Get "httpx://:9121/metrics2": unsupported protocol scheme "httpx".
[root@bb074911acd2 /]# pmm-admin add external --listen-port=9122 --scheme='http' --metrics-path='metrics2' --group="redis" --service-name="my-redis24" --metrics-mode=push
Connection check failed: Get "http://:9122/metrics2": dial tcp :9122: connect: connection refused.
[root@bb074911acd2 /]# pmm-admin add external --listen-port=9122 --scheme='http' --metrics-path='metrics2' --group="redis" --service-name="my-redis24" --metrics-mode=pull
Connection check failed: Get "http://172.20.0.4:9122/metrics2": dial tcp 172.20.0.4:9122: connect: connection refused.
[root@bb074911acd2 /]# pmm-admin add external --listen-port=9121 --scheme='httpx' --metrics-path='metrics2' --group="redis" --service-name="my-redis24" --metrics-mode=pull
Connection check failed: Get "httpx://172.20.0.4:9121/metrics2": unsupported protocol scheme "httpx".
[root@bb074911acd2 /]#
```

### CODE REVIEW

**Comment by AlekSi** [comment]
@askomorokhov please resolve merge conflicts

File: services/management/external.go [hunk 1]
```diff
@@ -18,10 +18,18 @@ package management
...
 	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
 
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
```
Comment (BupycHuk) [inline services/management/external.go:30]: let's remove it

**Review by JiriCtvrtka — APPROVED** [review]

File: services/management/external.go [hunk 2]
```diff
@@ -111,6 +119,18 @@ func (e *ExternalService) AddExternal(ctx context.Context, req *managementpb.Add
 			return err
 		}
 
+		if !req.SkipConnectionChecks {
+			if isPushMode(req.MetricsMode) {
+				if err = e.registry.CheckConnectionToService(ctx, tx.Querier, service, row); err != nil {
+					return err
+				}
+			} else {
+				if err = e.checkPullConnectionToService(ctx, tx.Querier, service, row); err != nil {
```
Comment (BupycHuk) [inline services/management/external.go:128]: BTW, to not duplicate code in pmm-managed and pmm-agent we can send CheckConnectionRequest to pmm-agent on pmm-server. What do you think?

Comment (askomorokhov) [reply services/management/external.go:128]: On, the first look sounds crazy, but let's try :)

**Review by percona-csalguero — APPROVED** [review]

File: services/management/external.go (hunk 2 above)
Comment (BupycHuk) [reply services/management/external.go:128]: For me too :-D

Comment (askomorokhov) [reply services/management/external.go]: Changed

File: services/agents/registry.go [hunk 3]
```diff
@@ -762,6 +765,10 @@ func (r *Registry) CheckConnectionToService(ctx context.Context, q *reform.Queri
 	}()
 
 	pmmAgentID := pointer.GetString(agent.PMMAgentID)
+	if !agent.PushMetrics && service.ServiceType == models.ExternalServiceType {
+		pmmAgentID = "pmm-server"
```
Comment (BupycHuk) [inline services/agents/registry.go:769]:
```suggestion
		pmmAgentID = models.PMMServerAgentID
```

File: services/agents/registry.go [hunk 4]
```diff
@@ -799,6 +806,28 @@ func (r *Registry) CheckConnectionToService(ctx context.Context, q *reform.Queri
 			Dsn:     agent.DSN(service, 2*time.Second, "", nil),
 			Timeout: ptypes.DurationProto(3 * time.Second),
 		}
+	case models.ExternalServiceType:
+		dsn := agent.DSN(service, 2*time.Second, "", nil)
+		if !agent.PushMetrics {
```
Comment (BupycHuk) [inline services/agents/registry.go:811]: I think all this stuff should be in agent_model.go

File: models/agent_model.go [hunk 5]
```diff
@@ -358,6 +358,24 @@ func (s *Agent) DSN(service *Service, dialTimeout time.Duration, database string
...
 		return u.String()
 
+	case ExternalExporterType:
+		scheme := pointer.GetString(s.MetricsScheme)
+		path := pointer.GetString(s.MetricsPath)
+		listenPort := int(pointer.GetUint16(s.ListenPort))
+		address := net.JoinHostPort(host, strconv.Itoa(listenPort))
+		u := &url.URL{
+			Scheme: scheme,
+			Host:   address,
```
Comment (BupycHuk) [inline models/agent_model.go:368]: exporter and service may be on different nodes.

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: services/agents/registry.go (hunk 3 above)
Comment (askomorokhov) [reply services/agents/registry.go]: Fixed.

File: services/agents/registry.go (hunk 4 above)
Comment (askomorokhov) [reply services/agents/registry.go]: After discussions extracted to another method.

File: models/agent_model.go [hunk 6]
```diff
@@ -357,12 +357,42 @@ func (s *Agent) DSN(service *Service, dialTimeout time.Duration, database string
...
 }
 
+func (s *Agent) ExporterURL(q *reform.Querier) (string, error) {
+	scheme := pointer.GetString(s.MetricsScheme)
+	path := pointer.GetString(s.MetricsPath)
+	listenPort := int(pointer.GetUint16(s.ListenPort))
+	username := pointer.GetString(s.Username)
+	password := pointer.GetString(s.Password)
+
+	host := ""
```
Comment (BupycHuk) [inline models/agent_model.go:372]: let's use 127.0.0.1

File: models/agent_model.go [hunk 7]
```diff
@@ -357,12 +357,42 @@ func (s *Agent) DSN(service *Service, dialTimeout time.Duration, database string
 			u.User = url.User(username)
 		}
 		return u.String()
-
 	default:
 		panic(fmt.Errorf("unhandled AgentType %q", s.AgentType))
 	}
 }
 
+func (s *Agent) ExporterURL(q *reform.Querier) (string, error) {
```
Comment (BupycHuk) [inline models/agent_model.go:365]: Can we write a few tests here?

File: models/agent_model.go (hunk 6 above)
Comment (askomorokhov) [reply models/agent_model.go]: Changed.

File: models/agent_model.go (hunk 7 above)
Comment (askomorokhov) [reply models/agent_model.go:365]: Added.

**Review by percona-csalguero — APPROVED** [review]

**Review by BupycHuk — APPROVED** [review]

**Review by JiriCtvrtka — APPROVED** [review]

**Pull request merged**